usl_pipeline/cloud_functions/main.py

```diff
@@ -34,44 +35,93 @@ class FeatureMetadata:
     elevation_max: Optional[float] = None
 
 
+def _retry_and_report_errors(
+    func: Callable[[functions_framework.CloudEvent], None]
+) -> Callable[[functions_framework.CloudEvent], None]:
+    """Decorator which adds error handling and retries to a cloud function."""
+
+    @functools.wraps(func)
+    def wrapper(cloud_event: functions_framework.CloudEvent) -> None:
+        logging.basicConfig(level=logging.WARN)
+
+        # To handle retries, check the time created and return if the event is too old.
+        # GCP retries until the function returns successfully, so the pattern is to
+        # report exceptions to have the function retried and return normally to stop
+        # retries. https://cloud.google.com/functions/docs/bestpractices/retries
+        event_time = datetime.datetime.fromisoformat(cloud_event.data["timeCreated"])
+        event_age = (
+            datetime.datetime.now(datetime.timezone.utc) - event_time
+        ).total_seconds()
+        if event_age > _MAX_RETRY_SECONDS:
+            logging.error(
+                "Dropped event id %s source %s after %s seconds",
+                cloud_event["id"],
+                cloud_event["source"],
+                _MAX_RETRY_SECONDS,
+            )
+            return
+
+        # Catch exceptions and report them with GCP error reporting.
+        try:
+            func(cloud_event)
+        except Exception as exc:  # noqa
+            # Log the exception to display in Cloud Function logs.
+            logging.exception(exc)
+            # Report the error in GCP Error Reporter.
+            error_reporting.Client().report_exception()
+            # Write the error to the metastore.
+            _write_chunk_metastore_error(cloud_event.data["name"], str(exc))
+
+    return wrapper
+
+
+@functions_framework.cloud_event
+@_retry_and_report_errors
+def write_study_area_metadata(cloud_event: functions_framework.CloudEvent) -> None:
+    """Writes metadata for a study area on upload.
+
+    This function is triggered when files containing raw geo data for an entire study
+    area are uploaded to GCP. It writes an entry to the metastore for the new study
+    area. It also calculates min & max values needed for feature scaling of the relevant
+    files.
+    """
+    file_name = pathlib.PurePosixPath(cloud_event.data["name"])
+    if file_name.name == "elevation.tif":
```

Comment:
I feel like "elevation.tif" deserves to be declared as a constant (maybe even in some shared file?) and many parties might be using this file name here and there. And other expected file names (like buildings.txt, green_areas.txt and so on) will be added there too once we get to that logic.

Reply:
ibid